### PR TITLE
chore: reorder pre-commit & pyupgrade --py39

### PR DESCRIPTION
**Pull request by MindTooth**
I have a complaint from MyPy:

```
src/core/repository/repository.py:83: error: Function "core.repository.repository.AbstractProjectRepository.list" is not valid as a type  [valid-type]
src/core/repository/repository.py:83: note: Perhaps you need "Callable[...]" or a callback protocol?
Found 1 error in 1 file (checked 1 source file)
```

```
❯ nina --dev
Traceback (most recent call last):
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/bin/nina", line 5, in <module>
    from nina.run import main
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/nina/run.py", line 9, in <module>
    from core.main import main as core_main  # type: ignore
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/main.py", line 14, in <module>
    import core.api
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/api/__init__.py", line 3, in <module>
    from .api import *
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/api/api.py", line 29, in <module>
    from core import model, services
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/services.py", line 24, in <module>
    from core.repository import SqlAlchemyProjectRepository as ProjectRepository
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/repository/__init__.py", line 3, in <module>
    from .repository import *
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/repository/repository.py", line 23, in <module>
    class AbstractProjectRepository(abc.ABC):
  File "/Users/birgerjn/Developer/fish-code/.direnv/python-3.9.6/lib/python3.9/site-packages/core/repository/repository.py", line 83, in AbstractProjectRepository
    def _list(self) -> list[model.Project]:  # pragma: no cover
TypeError: 'function' object is not subscriptable
```

Not sure what this tells me.

I see that it's not implemented.  Warranted a removal?

### CODE REVIEW

**Comment by tomrtk** [comment]
To solve the typeerror at runtime we need `from __future__ import annotations` in the file or python tries to index the function `list`.

> I have a complaint from MyPy:
> ...
> Not sure what this tells me.
> 
> I see that it's not implemented. Warranted a removal?

The  thought of the abstract class was to define the api for an implementation of the repository.
to solve the typing issue we can define a [`Protocol`][1] of the minimal signature a repository needs.

something like this seems to work, I can make a PR if you want.
```diff
diff --git a/src/core/repository/repository.py b/src/core/repository/repository.py
index d63b66c..17ff387 100644
--- a/src/core/repository/repository.py
+++ b/src/core/repository/repository.py
@@ -5,7 +5,7 @@ abstract class defines the methods all repositories needs to implement.
 """
 import abc
 import logging
-from typing import List, Optional
+from typing import List, Optional, Protocol
 
 from sqlalchemy.orm.session import Session
 
@@ -20,7 +20,21 @@ class NotFound(Exception):
     pass
 
 
-class AbstractProjectRepository(abc.ABC):
+class _ProjectRepositroy(Protocol):
+    def _add(self, project: model.Project) -> model.Project:
+        ...
+
+    def _save(self) -> None:
+        ...
+
+    def _get(self, reference: int) -> Optional[model.Project]:
+        ...
+
+    def _list(self) -> list[model.Project]:
+        ...
+
+
+class AbstractProjectRepository(abc.ABC, _ProjectRepositroy):
     """Abstract project repository defining minimum API of repository."""
 
     def __init__(self) -> None:
@@ -65,24 +79,6 @@ class AbstractProjectRepository(abc.ABC):
         """Get a list off all Projects in repository."""
         return self._list()
 
-    @abc.abstractmethod
-    def _add(self, project: model.Project) -> model.Project:  # pragma: no cover
-        raise NotImplementedError
-
-    @abc.abstractmethod
-    def _save(self) -> None:  # pragma: no cover
-        raise NotImplementedError
-
-    @abc.abstractmethod
-    def _get(
-        self, reference: int
-    ) -> Optional[model.Project]:  # pragma: no cover
-        raise NotImplementedError
-
-    @abc.abstractmethod
-    def _list(self) -> list[model.Project]:  # pragma: no cover
-        raise NotImplementedError
-
     def __len__(self) -> int:
         """Get number of projects in repository.
```

Also if the logging and `__len__` is move to `SqlAlchemyProjectRepository` the abstract class can be removed and the protocol defines the api.

[1]: https://mypy.readthedocs.io/en/stable/protocols.html?highlight=protocal#protocols-and-structural-subtyping

**Comment by MindTooth** [comment]
Feel free to open a pull request. 😁  Thanks for the explanation.

**Comment by MindTooth** [comment]
Did a rebase and force-push to remove my failed attempt at solving the issue.  Seems to retain the author.  Nice work!

File: src/core/model.py [hunk 1]
```diff
@@ -4,21 +4,12 @@
...
-    DefaultDict,
-    Dict,
-    Generator,
-    List,
-    Optional,
-    Set,
-    Tuple,
-    Union,
-)
+from typing import Any, DefaultDict, Dict, List, Optional, Set, Tuple, Union
```
Comment (tomrtk) [inline src/core/model.py:12]: Looks like there is a few imports that is now unused in this and other files. I see we ignore `F401` in flake8 config, so it is not caught by checks. We can maybe look at that as a separate issue.

**Review by tomrtk — APPROVED** [review]

File: src/core/model.py (hunk 1 above)
Comment (MindTooth) [reply src/core/model.py:12]: Easy. 😅

**Pull request merged**